napari/layers/_multiscale_data.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from ._data_protocols import LayerDataProtocol, assert_protocol
+from .utils.layer_utils import compute_multiscale_level_and_corners
+
+
+class MultiScaleData(Sequence[LayerDataProtocol], LayerDataProtocol):
+    """Wrapper for multiscale data, to provide consistent API.
+
+    :class:`LayerDataProtocol` is the subset of the python Array API that we
+    expect array-likes to provide.  Multiscale data is just a sequence of
+    array-likes (providing, e.g. `shape`, `dtype`, `__getitem__`).
+
+    Parameters
+    ----------
+    data : Sequence[LayerDataProtocol]
+        Levels of multiscale data, from larger to smaller.
+    max_size : Sequence[int], optional
+        Maximum size of a displayed tile in pixels, by default`data[-1].shape`
+
+    Raises
+    ------
+    ValueError
+        If `data` is empty or is not a list, tuple, or ndarray.
+    TypeError
+        If any of the items in `data` don't provide `LayerDataProtocol`.
+    """
+
+    def __init__(
+        self,
+        data: Sequence[LayerDataProtocol],
+        max_size: Optional[Sequence[int]] = None,
+    ) -> None:
+        self._data: List[LayerDataProtocol] = list(data)
+        if not self._data:
+            raise ValueError("Multiscale data must be a (non-empty) sequence")
+        for d in self._data:
+            assert_protocol(d)
+
+        self.max_size = self._data[-1].shape if max_size is None else max_size
+        self.downsample_factors = (
+            np.array([d.shape for d in data]) / data[0].shape
+        )
+
+    @property
+    def dtype(self) -> np.dtype:
+        """Return dtype of the first scale.."""
+        return self._data[0].dtype
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Shape of multiscale is just the biggest shape."""
+        return self._data[0].shape
+
+    @property
+    def shapes(self) -> Tuple[Tuple[int, ...], ...]:
+        """Tuple shapes for all scales."""
+        return tuple(im.shape for im in self._data)
+
+    def __getitem__(  # type: ignore [override]
+        self, key: Union[int, Tuple[slice, ...]]
```

Comment:
The signature of `__getitem__` is less specific than in `LayerDataProtocol` - should we consider supporting fewer types in `LayerDataProtocol` (i.e. the ones here)?

Reply:
Yeah, this points to just how difficult the case of multiscale data is.  For better or worse, multiscale data _isn't_ just like an array... and it will require a lot more discussion to fully flesh out what it could look like.  You'll note here that inasumuch as this protocol doesn't fully implement the `LayerDataProtocol` (which also acknowledges that sometimes one will try to index an array-like with another array-like), we actually raise a `NotImplementedError`.

In other words, we're not trying to "fully tackle" this (very big) challenge in this PR.  We're just:
1. acknowledging (in code) that it's a problem
2. introducing a protocol (`LayerDataProtocol`) that is at least a place for us to document the methods and signatures that we intend to use internally as we deal `with layer.data` (rather than just saying "give us something like an array and it will probably work")
3. Taking a first step to make multiscale data not totally choke with simple things like `data.dtype` and `data.shape`... and `__getitem__` here is a first step to provide a sane default for indexing into this "list-of-arrays" with anything other than a single integer.  But yeah, it's not completely implementing our `LayerDataProtocol` until we can define what that behavior should be.  And so we raise a `NotImplementedError`

Reply:
> should we consider supporting fewer types in LayerDataProtocol (i.e. the ones here)?

we can't do that because we already know that we use fancy indexing in some places internally... so it wouldn't be "true"

Reply:
I see. So by implication, `MultiScaleData` won't work in general (as a `LayerDataProtocol`), and the typing is explicitly acknowledging that. Agreed that is good progress and not everything has to be address in this PR.

Reply:
Also, I guess the `# type: ignore [override]` is actively silencing type checkers? When adding those kind of things, it's always good to add an explanatory comment to explain why you're ignoring it. I think you could just copy a concise version of the reason you have above.